doc/man/slogencrypt.1.in
.TH "slogencrypt"1 "" "" slogencrypt
.PP
.SS "Name"
.PP
slogencrypt — Encrypt existing plain text log files using the syslog-ng secure logging environment.
.SS "Synopsis"
.PP
slogencrypt [options] [arguments]
.SS "Description"
.PP
The slogencrypt utility is used to encrypt plain text log file using an existing secure logging environment. Using this utility, log files obtained from a previous installation of syslog-ng or another logging system can be transferred to a secure logging environment. The order of the log entries is preserved. Encrypting plain text log files using an existing secure logging environment, requires the current encryption key to be supplied in order to preserve consistency.
.PP
General call sequence: slogencrypt -k &lt;key file&gt; -m &lt;MAC file&gt; &lt;new key file&gt; &lt;new MAC file&gt; &lt;plain text log&gt; &lt;output file&gt; [counter]
.SS "Arguments"
.RS
.IP * 2
&lt;new key file&gt;
.IP
The file that will contain the new current encryption key after successful encryption.
.IP *
&lt;new MAC file&gt;
.IP
The file receiving the new current message authentication code (MAC) of the secure encrypted destination after encryption. In case an existing file is supplied, new entries will be appended.
.IP *
&lt;input log file&gt;
.IP
The plain text log file that will be encrypted using the secure logging environment.
.IP *
&lt;output log file&gt;
.IP
The file that will contain the encrypted log entries from the supplied plain text log file after encryption.
.IP *
counter
.IP
The current log entry counter of the secure encrypted destination after encryption. This is required if the log entries to be encrypted will be appended to an existing secure encrypted destination.
.RE
.SS "Options"
.RS
.IP * 2
--key-file or -k
.IP
The current host key from the system where the encryption will be performed.
.IP *
--mac-file or -m
.IP
The current MAC file from the system where the encryption will be performed.
.IP *
--help or -h
.IP
Display a help message.
.RE
.SS "Files"
.PP
/usr/bin/slogencrypt
.PP
/etc/syslog-ng.conf
.SS "See also"
.PP
\fBsyslog-ng.conf.5\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/syslog-ng.conf.5/\fP)
.PP
\fBsyslog-ng.8\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/syslog-ng.8/\fP)
.SS "Getting help"
.RS
.IP * 2
The up-to-date documentation of syslog-ng is available on the \fBAxoSyslog documentation site\fP (\fIhttps://axoflow.com/docs/axosyslog-core/\fP).
.IP *
For news and notifications about syslog-ng, visit the \fBAxoflow blog\fP (\fIhttps://axoflow.com/blog/\fP).
.IP *
If you want to contact the developers directly to help with problems or report issues, contact us on \fBDiscord\fP (\fIhttps://discord.gg/583Z4wjem2\fP) or \fBGitHub\fP (\fIhttps://github.com/syslog-ng/syslog-ng/issues/\fP).
.RE
.PP
This manual page is maintained by \fBAxoflow\fP (\fIhttps://axoflow.com/\fP)
Last modified November 24, 2023: \fBMinor manpage fix (2f0a144)\fP (\fIhttps://github.com/axoflow/axosyslog-core-docs/commit/2f0a144ec391a3182da3af2f2cc136042d5608e9\fP)

NEWS
3.5.6
        Mon,  4 Aug 2014 13:39:42 +0200

        This is the sixth bug-fix release for the 3.5.x series.
        Upgrading from earlier versions is highly recommended, as the
        changes in this release are very small, yet, also very
        important for most platforms and workloads.

        Bugfixes
        ========

        * A major memory leak was fixed in the value-pairs framework,
          which affects $(format-json), MongoDB, AMQP and more. The
          leak was supposed to be fixed in 3.5.5, but due to a merging
          mistake, it was missed.

        * The configure script now detects Linux capabilities
          properly, there is no need to use --disable-linux-caps on
          non-Linux platforms anymore.

        * The pdbtool merge command will now generate version 4
          patterndb files, instead of version 3.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Fabien Wernli
        Gergely Nagy
        Peter Czanik

3.5.5
        Mon, 21 Jul 2014 17:00:00 +0200

        This is the fifth bug-fix release for the 3.5.x series.
        Starting with this release, the 3.5 branch replaces 3.4 as the
        current "stable" version. Upgrading from earlier versions -
        and especially earlier branches - is highly recommended.

        New features
        ============

        * The facility() filter now supports ranges, allowing one to
          capture a range of facilities: facility(local0..local6).

        * The $(format-json) template function now supports SEQNUM.

        * The upstart script in contrib has been improved to allow the
          user to override certain settings (such as the configuration
          file, and the pidfile), and allow extra options to be
          specified.

        Bugfixes
        ========

        * Major memory leaks have been fixed in various parts of
          syslog-ng, including but not limited to network and file
          destinations and sources, value-pairs (thus affecting
          mongodb, $(format-json), and others), template handling and
          rewrite rules.

          Some of these leaks only happened during reload, some
          leaked on every message. They have now been plugged.

        * The value-pairs framework has seen a lot of updates, too:

          - It now honors time zone settings, so everything that uses
            value-pairs, will correctly take time zone settings into
            account.

          - Using globs in the key() option of value pairs will now
            also search macros for matches (thus key("D*") will match
            DATE too).

          - The syslog-proto (RFC5424) scope now correctly includes
            SDATA.

          - A bug that resulted in hierarchical elements being
            misparsed and made to appear in random places have been
            fixed.

        * The body() and subject() options of the SMTP destinations
          are now mandatory, and a double-free was also fixed in the
          driver.

        * Minor fixes were made to the $(env) template function, so
          that when looking up an empty environment variable, we will
          no longer trigger a glib warning. A misleading error message
          was also corrected.

        * The pdbtool patternize tool will now output version 4 PDB
          XML.

        * The configure script will not try to configure the AMQP
          module if no suitable version of python is installed. This
          avoids a difficult to diagnose error message during
          compilation.

        Developer news
        ==============

        * When compiled with internal ivykis, we will now install its
          headers too, along with our own. This is to make it easier
          for the Incubator to be compiled against a syslog-ng with
          embedded ivykis.

        * Plugin symbols are now loaded differently, in a way that
          lets plugins load further plugins, and use already loaded
          symbols. This is required for the various scripting
          destinations in the Incubator.

        * Starting with this release, we are installing the testing
          helpers as a convenience library (available under the
          syslog-ng-test pkg-config module). This allows third party
          projects to use our helpers.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Balazs Scheidler <bazsi@balabit.hu>
        Daniel Gados
        Fabien Wernli
        Gergely Nagy <algernon@balabit.hu>
        Laszlo Budai <lbudai@balabit.hu>
        Laszlo Meszaros <lmesz@balabit.hu>
        Nick Alcock <nix@esperi.org.uk>
        Peter Czanik <czanik@balabit.hu>
        Peter Gyongyosi <gyp@balabit.hu>
        Sebastien Badia <seb@sebian.fr>
        Tibor Benke <tibor.benke@balabit.com>
        Viktor Juhasz <viktor.juhasz@balabit.com>
        Viktor Tusa <tusa@balabit.hu>
        Xufeng Zhang <xufeng.zhang@windriver.com>
        Zoltan Pallagi <pzolee@balabit.hu>

3.5.4.1
        Thu, 13 Mar 2014 10:46:07 +0100

        This is the fourth bug-fix release for the 3.5.x series, with
        a single fix for 32-bit platforms. Upgrading to this version
        is highly recommended for people using 32-bit platforms,
        others are not affected.


modules/secure-logging/slogencrypt/slogencrypt.c
/*
 * Copyright (c) 2019 Airbus Commercial Aircraft
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <glib.h>

#include "messages.h"
#include "slog.h"
#include "compat/string.h"

// Arguments and options
static char *hostkey = NULL;
static char *newhostKey = NULL;
static char *inputMAC = NULL;
static char *outputMAC = NULL;
static char *inputlog = NULL;
static char *outputlog = NULL;
static guint64 bufSize = DEF_BUF_SIZE;

int main(int argc, char *argv[])
{
  int index = 0;
  SLogOptions options[] =
  {
    { "key-file", 'k', "Current host key file", "FILE", NULL },
    { "mac-file", 'm', "Current MAC file", "FILE", NULL },
    { NULL }
  };

  GOptionEntry entries[] =
  {
    { options[0].longname, options[0].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[0].description, options[0].type },
    { options[1].longname, options[1].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[1].description, options[1].type },
    { NULL }
  };

  GError *error = NULL;
  GOptionContext *context =
    g_option_context_new("NEWKEY NEWMAC INPUTLOG OUTPUTLOG [COUNTER] - Encrypt log files using secure logging");
  GOptionGroup *group = g_option_group_new("Basic options", "Basic log encryption options", "basic", &options, NULL);

  g_option_group_add_entries(group, entries);
  g_option_context_set_main_group(context, group);

  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      GString *errorMsg = g_string_new(error->message);

      return slog_usage(context, group, errorMsg);
    }

  if(argc < 5 || argc > 6)
    {
      return slog_usage(context, group, NULL);
    }

  // Initialize internal messaging
  msg_init(TRUE);

  char key[KEY_LENGTH];
  char mac[CMAC_LENGTH];

  // Assign option arguments
  index = 0;
  hostkey = options[index].arg;
  index++;
  inputMAC = options[index].arg;

  // Input and output file arguments
  index = 1;
  newhostKey = argv[index];
  index++;
  if(newhostKey == NULL)
    {
      return slog_usage(context, group, NULL);
    }

  outputMAC = argv[index];
  index++;
  if(outputMAC == NULL)
    {
      return slog_usage(context, group, NULL);
    }

  inputlog = argv[index];
  index++;
  if(!g_file_test(inputlog, G_FILE_TEST_IS_REGULAR))
    {
      GString *errorMsg = g_string_new(FILE_ERROR);
      g_string_append(errorMsg, inputlog);
      return slog_usage(context, group, errorMsg);
    }

  outputlog = argv[index];
  index++;
  if(outputlog == NULL)
    {
      return slog_usage(context, group, NULL);
    }

  // Read key and counter
  guint64 counter;
  int ret = readKey(key, &counter, hostkey);
  if (ret!=1)
    {
      msg_error("[SLOG] ERROR: Unable to read host key", evt_tag_str("file", hostkey));
      return -1;
    }

  // Buffer size arguments if applicable
  if (argc == 6)
    {
      sscanf(argv[index], "%"G_GUINT64_FORMAT, &bufSize);

      if(bufSize <= MIN_BUF_SIZE || bufSize > MAX_BUF_SIZE)
        {
          msg_error("[SLOG] ERROR: Invalid buffer size.", evt_tag_int("Size", bufSize),
                    evt_tag_int("Minimum buffer size", MIN_BUF_SIZE), evt_tag_int("Maximum buffer size", MAX_BUF_SIZE));
          g_option_context_free(context);
          return 0;
        }
    }

  // Open input file
  FILE *inputFile = fopen(inputlog, "r");
  if (inputFile == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to open input log file", evt_tag_str("file", inputlog));
      return -1;
    }

  // Open output file
  FILE *outputFile = fopen(outputlog, "w");
  if (outputFile == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to open output log file", evt_tag_str("file", outputlog));
      return -1;
    }

  // Read MAC (if possible)
  if (readBigMAC(inputMAC, mac)==0)
    {
      msg_warning("[SLOG] ERROR: Unable to open input MAC file", evt_tag_str("file", inputMAC));
    }

  size_t readLen = 0;
  char *line = NULL;

  msg_info("Importing data...");

  // Parse data
  while(getline(&line, &readLen, inputFile)!=-1)
    {
      char outputmacdata[CMAC_LENGTH];

      GString *result = g_string_new(NULL);
      GString *inputGString = g_string_new(line);

      // Remove trailing '\n' from string
      g_string_truncate(inputGString, (inputGString->len)-1);

      gsize outputmacdata_capacity = G_N_ELEMENTS(outputmacdata);
      sLogEntry(counter, inputGString, (unsigned char *)key, (unsigned char *)mac, result, (unsigned char *)outputmacdata,
                outputmacdata_capacity);

      fprintf(outputFile, "%s\n", result->str);

      // Update keys, MAC, etc
      memcpy(mac, outputmacdata, CMAC_LENGTH);
      evolveKey((unsigned char *)key);
      counter++;

      readLen = 0;
      free(line);
      g_string_free(result, TRUE);
      g_string_free(inputGString, TRUE);
      line = NULL;
    }


  // Write whole log MAC
  if (writeBigMAC(outputMAC, mac)==0)
    {
      msg_error("Problem with output MAC file");
      return -1;
    }

  // Write key
  ret = writeKey(key, counter, newhostKey);
  if (ret!=1)
    {
      msg_error("[SLOG] ERROR: Unable to write new host key", evt_tag_str("file", outputlog));
      return -1;
    }

  fclose(outputFile);

  msg_info("All data successfully imported.");

  return 1;
}
